qiskit/circuit/parameter.py

```diff
@@ -52,34 +54,27 @@ class Parameter(ParameterExpression):
 
     __slots__ = ("_name", "_uuid", "_hash")
 
-    def __new__(cls, name, uuid=None):  # pylint: disable=unused-argument
-        # Parameter relies on self._uuid being set prior to other attributes
-        # (e.g. symbol_map) which may depend on self._uuid for Parameter's hash
-        # or __eq__ functions.
-        obj = object.__new__(cls)
-
-        if uuid is None:
-            obj._uuid = uuid4()
-        else:
-            obj._uuid = uuid
-
-        obj._hash = hash(obj._uuid)
-        return obj
+    # This `__init__` does not call the super init, because we can't construct the
+    # `_parameter_symbols` dictionary we need to pass to it before we're entirely initialised
+    # anyway, because `ParameterExpression` depends heavily on the structure of `Parameter`.
 
-    def __getnewargs__(self):
-        # Unpickling won't in general call __init__ but will always call
-        # __new__. Specify arguments to be passed to __new__ when unpickling.
-
-        return (self.name, self._uuid)
-
-    def __init__(self, name: str):
+    def __init__(
+        self, name: str, *, uuid: UUID | None = None
+    ):  # pylint: disable=super-init-not-called
         """Create a new named :class:`Parameter`.
 
         Args:
             name: name of the ``Parameter``, used for visual representation. This can
                 be any unicode string, e.g. "ϕ".
+            uuid: For advanced usage only.  Override the UUID of this parameter, in order to make it
+                compare equal to some other parameter object.  By default, two parameters with the
+                same name do not compare equal to help catch shadowing bugs when two circuits
+                containing the same named parameters are spurious combined.  Setting the ``uuid``
+                field when creating two parameters to the same thing (along with the same name)
+                allows them to be equal.  This is useful during serialization and deserialization.
         """
         self._name = name
```

Comment:
Might be unrelated to this PR, but why is the `_name` attribute needed? Isn't that information already included in `_symbol_expr`?

Reply:
Off the top of my head, I think you're completely right and there's no real problems with dropping this field to reduce a pointer's worth of memory usage per parameter, and slightly reduce the hashing work we need to do.

Because there's a (minor) risk that we break something by inferring `name` from the sympy/symegine symbol rather than storing it ourselves, perhaps let's try that in a follow-up PR so we can isolate it if it does happen to cause bugs without losing this bugfix.

Reply:
Ok I created https://github.com/Qiskit/qiskit/issues/10880